src/api/layer.js

```diff
@@ -125,24 +125,15 @@ export default class Layer {
      * @instance
      * @api
      */
-    async setSource(source) {
-        this._checkSource(source);
-        const style = this._style;
-        if (style) {
-            var metadata = await source.requestMetadata(style);
-        }
-        if (this._style !== style) {
-            throw new Error('A style change was made before the metadata was retrieved, therefore, metadata is stale and it cannot be longer consumed');
-        }
-        this.metadata = metadata;
-        source.bindLayer(this._onDataframeAdded.bind(this), this._onDataFrameRemoved.bind(this), this._onDataLoaded.bind(this));
-        if (this._source !== source) {
-            this._freeSource();
+    setSource(source) {
+        if (this._source === source) {
+            return;
         }
+        this._checkSource(source);
+        this._freeSource();
         this._source = source;
```

Comment:
This refactor looks very, very, dangerous to me. You are setting the object source to the new source before waiting for checking the metadata against the style.

Reply:
Whats the possible problem?

Reply:
Leaving the layer object in an invalid state.